lib/vsc/modules/cache.py

```diff
@@ -0,0 +1,290 @@
+            LOGGER.debug("Skipping spiderT key %s", mpath)
+            continue
+
+        if mpath in mpmap:
+            modulepaths.append(mpath)
+        else:
+            LOGGER.debug("Skipping modulepath %s not in modulepath map %s", mpath, mpmap)
+
+    modulepaths.sort()
+    LOGGER.debug("Found pre-sorted modulepaths %s", modulepaths)
+
+    # sort them
+    #   very trivial sort based on EXTRA_MODULEPATHS from CLUSTER_DATA
+    #   we do not assume complicated nesting
+    #   just search all EXTRA_MODULEPATHS last
+    #      they are listed in prepend order, but default/most sensible one is prepended last
+    #      so they are reversed: the first extra has to be moved as far as possible
+    for extras in [x.get('EXTRA_MODULEPATHS', [])[::-1] for x in CLUSTER_DATA.values()]:
+        for extra in extras:
+            # move to the end
+            LOGGER.debug("Moving EXTRA_MODULEPATH %s to the end", extra)
+            try:
+                modulepaths.remove(extra)
+                modulepaths.append(extra)
+            except ValueError as err:
+                LOGGER.error("Did not find EXTRA_MODULEPATH %s in modulespaths %s: %s", extra, modulepaths, err)
+
+    LOGGER.debug("Sorted modulepaths %s", modulepaths)
+    return modulepaths
+
+
+def sort_recent_versions(versions):
+    """Sort versions using LooseVersion, most recent first"""
+    return sorted(versions, key=LooseVersion, reverse=True)
+
+
+def software_map(spiderT, mpmap):
+    """Create a software map with what software has which versions on which cluster"""
+    modulepaths = sort_modulepaths(spiderT, mpmap)
+
+    softmap = {}
+    for mpath in modulepaths:
+        LOGGER.debug("Processing modulepath %s", mpath)
+        clusters = mpmap[mpath]
+        for name, namedata in spiderT[mpath].items():
+            soft = softmap.setdefault(name, {})
+            # all versions of this software in current modulepath
+            mpversions = []
+            for fullname, fulldata in namedata['fileT'].items():
+                version = fulldata['Version']
+                # sanity check
+                txt = "for modulepath %s name %s fullname %s: %s" % (mpath, name, fullname, fulldata)
+                if version != fulldata['canonical']:
+                    LOGGER.raiseException("Version != canonical " + txt)
+                if fullname != "%s/%s" % (name, version):
+                    LOGGER.raiseException("fullname != name/version " + txt)
+
+                mpversions.append(version)
+                softversion = soft.setdefault(version, [])
+                softversion.extend(clusters)
+
+            # determine default
+            #   the default is per clusters (actually per modulepath)
+            default = None
+            defaultdata = namedata['defaultT']
+
+            if defaultdata:
+                value = defaultdata['value']
+                if value:
+                    # default has full name, we only need the versionxs
```

Comment:
I love how creative you get with typos ;) (`versionxs` -> `versions`)

Reply:
fixed